addons/web/static/tests/boot_tests.js

```diff
@@ -1,18 +1,66 @@
-(function () {
+(async function () {
 "use strict";
 
-odoo.__DEBUG__.didLogInfo.then(function () {
-    var modulesInfo = odoo.__DEBUG__.jsModules;
+    // We define here a test asserting that no module failed to be loaded. The
+    // module system (see boot.js) waits 10 seconds before logging that some
+    // modules couldn't be loaded, when it happens. The test below is really
+    // important as it will detect testing modules that can't be loaded, for
+    // instance because one of their dependencies is missing. Without it, we
+    // could have tests that aren't executed, without any notice.
+    //
+    // We have the following constraints around this test:
+    //  - we want it to be always executed (obviously)
+    //  - we need it to wait for the module system to complete the loading of
+    //    the modules before being executed (wait for the didLogInfo promise)
+    //  - we don't want it to slow down the whole suite
+    //  - we can't rely on the module system to define this test, as it is
+    //    partially testing it as well
+    //
+    // To achieve this, we ensure that this test is defined after (some of) the
+    // other tests. Almost all tests need module 'web.test_utils', which is
+    // asynchronous (it loads resources that are lazy-loaded outside the test
+    // environment). So here, we also wait for 'web.test_utils' (without using
+    // the module system) before defining our test, so that our test is defined
+    // after all those requiring 'web.test_utils'. That way:
+    //  - either the suite takes more than 10 seconds to execute, and once we
+    //    reach our test, the promise is already resolved and it can directly
+    //    check if some modules failed
+    //  - or the suite takes less than 10 seconds, and our test will wait for
+    //    the promise to be resolved.
+    // However, in any cases, it won't slow down the test suite longer than
+    // necessary.
+    let resolve;
+    const prom = new Promise(r => {
+        resolve = r;
+    });
+    const timeout = setTimeout(() => {
+        // module 'web.test_utils' must be loaded after a while, otherwise it
+        // could mean it has been renamed, and our test will never been executed
+        if (!odoo.__DEBUG__.services['web.test_utils']) {
+            throw new Error('Module "web.test_utils" could not be loaded');
```

Comment:
I feel like this exception inside a `setTimeout` is not the best idea, as nothing will catch it/I won't crash the main code, but I guess as long as there is a log in the console the test tools will consider it a fail?

I feel like this should be on a `Promise` with a `reject`, wrap the `setInterval` inside another Promise (and then no need to extract the `resolve`), and make the 2 promises race?

Reply:
the purpose of this throw is to detect if someone renames the module 'web.test_utils'. We have at least a suite that takes more than 5sec (the main one), so it will be detected. Without this, I can't be sure that the test below is ever defined.

Reply:
I see, this line is intended to work from other tests than the failing one. I was not seeing it from that point of view. Maybe just add this in the comment of the timeout, that it is intended to be caught only from a test that has both no dependency issue and that it is long enough (it's still a lot of conditions for it to go right, but it's still better than before).

Reply:
note that we're talking about the hypothesis where the module 'web.test_utils' has been renamed by someone that didn't grep in the code where it is called (or worse, rename *manually* by someone who don't know how to do a 'replace all'). I'll add a bit more explanations, but I think this part will most likely never be useful.